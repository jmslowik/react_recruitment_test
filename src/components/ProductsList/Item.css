.row {
  padding: 10px;
  border-bottom: 1px solid silver;
  list-style: none;
  display: flex;
}

.after {
  margin-left: auto;
}

.price {
  margin-left: auto;
}

.item {
  width: 60%;
  display: flex;
}