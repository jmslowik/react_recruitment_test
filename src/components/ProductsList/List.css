.container{
    width: 1000px;
    margin: 100px auto 0 auto;
}

.list {
    margin-bottom: 0;
    border-bottom-style: solid;
}

.list .row:last-child {
    border-bottom-style: none;
}

.sum {
    font-weight: bold;
}