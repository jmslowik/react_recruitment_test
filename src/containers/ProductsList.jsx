import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getAllAction } from '@/store/products/saga';
import {
  setAndVerifyProductQuantityAction,
  setAndVerifyProductQuantityDebounceAction,
} from '@/store/basket/saga';
import { productsListSelector } from '@/store/products/selectors';
import { basketSelector } from '@/store/basket/selectors';
import ProductsListComponent from '@/components/ProductsList';

const ProductsList = () => {
  const dispatch = useDispatch();

  const products = useSelector(productsListSelector);
  const basket = useSelector(basketSelector);

  const onChange = ({ pid, newValue, min }) => {
    dispatch(setAndVerifyProductQuantityDebounceAction({ pid, quantity: newValue, min }));
  };

  useEffect(() => {
    dispatch(getAllAction());
  }, [dispatch]);

  useEffect(() => {
    products.forEach(({ pid, min }) => {
      dispatch(setAndVerifyProductQuantityAction({ pid, quantity: min, min }));
    });
  }, [products, dispatch]);

  return (
    <ProductsListComponent
      list={products}
      quantities={basket}
      onChange={onChange}
    />
  );
};

export default ProductsList;
