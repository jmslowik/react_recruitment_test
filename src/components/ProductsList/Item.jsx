import React, { useContext } from 'react';
import LocaleContext from '@/contexts/LocaleContext';
import './Item.css';

const Item = ({
  name, price = 0, after,
}) => {
  const { locale, currency } = useContext(LocaleContext);

  return (
    <li className="row">
      <div className="item">
        {`${name}:`}
        <span className="price">
          {price.toLocaleString(locale, { style: 'currency', currency })}
        </span>

      </div>
      <div className="after">{after}</div>
    </li>
  );
};

export default Item;
