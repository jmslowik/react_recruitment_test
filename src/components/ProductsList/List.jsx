import React from 'react';
import './List.css';

const ProductsList = ({ children }) => (
  <div className="container">
    <h3>Lista produkt√≥w</h3>
    <ul>
      { children }
    </ul>
  </div>
);

export default ProductsList;
