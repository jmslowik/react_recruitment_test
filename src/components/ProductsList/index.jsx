import React from 'react';
import List from '@/components/ProductsList/List';
import Item from '@/components/ProductsList/Item';
import QuantityController from '@/components/QuantityController';
import { parsePrice, getSum, getTotalQuantity } from '@/utils';

const ProductsList = ({ list, quantities, onChange }) => {
  const totalPrice = getSum(list, quantities);
  const totalQuantity = getTotalQuantity(quantities);

  const Sum = () => (
    <Item
      name="Suma"
      price={totalPrice}
      after={<div>{`Obecnie masz ${totalQuantity} przedmiotów w koszyku`}</div>}
    />
  );

  return (
    <List name="Lista produktów" sum={<Sum />}>
      { list && list.map(({
        pid, name, price, min, max, isBlocked,
      }) => (
        quantities[pid] && (
        <Item
          key={pid}
          name={name}
          price={parsePrice(price)}
          after={(
            <QuantityController
              quantity={(quantities[pid] && quantities[pid].quantity) || min}
              min={min}
              max={max}
              isBlocked={isBlocked}
              onChange={(newValue) => onChange({ pid, newValue, min })}
            />
          )}
        />
        )
      ))}
    </List>
  );
};

export default ProductsList;
