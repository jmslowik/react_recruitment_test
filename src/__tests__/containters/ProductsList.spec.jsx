import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { dispatchCallsByTypeFilter } from '@/__tests__/__utils__';

import ProductsList from '@/containers/ProductsList';

const mockStore = configureStore([]);

describe('ProductsList container', () => {
  let store;
  const renderComponent = () => renderer.create(
    <Provider store={store}>
      <ProductsList />
    </Provider>,
  );
  const mountComponent = () => mount(
    <Provider store={store}>
      <ProductsList />
    </Provider>,
  );

  beforeEach(() => {
    store = mockStore({
      products: [
        {
          pid: 'pid-1', name: 'name 1', price: 1, isBlocked: true,
        },
        { pid: 'pid-2', name: 'name 2', price: 2 },
      ],
      basket: {
        'pid-1': { quantity: 1 },
        'pid-2': { quantity: 2 },
      },
    });

    store.dispatch = jest.fn();
  });

  it('should render with given state from Redux store', () => {
    const component = renderComponent();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch [PRODUCTS] GET_ALL on mount first', () => {
    mountComponent();
    const callsFilter = dispatchCallsByTypeFilter('[PRODUCTS] GET_ALL');

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenNthCalledWith(1, { type: '[PRODUCTS] GET_ALL', payload: {} });
    expect(store.dispatch.mock.calls.filter(callsFilter)).toHaveLength(1);
  });

  it('should dispatch [BASKET] INIT_PRODUCT_QUANTITIES on mount', () => {
    mountComponent();
    const type = '[BASKET] INIT_PRODUCT_QUANTITIES';

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenNthCalledWith(2, { type, payload: expect.anything() });
  });
});
