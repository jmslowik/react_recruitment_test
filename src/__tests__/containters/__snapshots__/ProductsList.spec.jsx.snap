// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList container should render with given state from Redux store 1`] = `
<div
  className="container"
>
  <h3>
    Lista produktów
  </h3>
  <ul
    className="list"
  >
    <li
      className="row"
    >
      <div
        className="item"
      >
        name 1:
        <span
          className="price"
        >
          1,00 zł
        </span>
      </div>
      <div
        className="after"
      >
        <div
          className="quantity-controller"
        >
          <div
            className="quantity-text"
          >
            Obecnie masz
             
            <span
              className="quantity"
            >
              1
            </span>
             
            sztuk produktu
          </div>
          <button
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            -
          </button>
          <button
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            +
          </button>
        </div>
      </div>
    </li>
    <li
      className="row"
    >
      <div
        className="item"
      >
        name 2:
        <span
          className="price"
        >
          2,00 zł
        </span>
      </div>
      <div
        className="after"
      >
        <div
          className="quantity-controller"
        >
          <div
            className="quantity-text"
          >
            Obecnie masz
             
            <span
              className="quantity"
            >
              2
            </span>
             
            sztuk produktu
          </div>
          <button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            -
          </button>
          <button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            +
          </button>
        </div>
      </div>
    </li>
  </ul>
  <ul
    className="sum"
  >
    <li
      className="row"
    >
      <div
        className="item"
      >
        Suma:
        <span
          className="price"
        >
          5,00 zł
        </span>
      </div>
      <div
        className="after"
      >
        <div>
          Obecnie masz 3 przedmiotów w koszyku
        </div>
      </div>
    </li>
  </ul>
</div>
`;
